sync/settings.go

```diff
@@ -0,0 +1,305 @@
+package sync
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/ghodss/yaml"
+
+	"github.com/safing/portbase/api"
+	"github.com/safing/portbase/config"
+	"github.com/safing/portmaster/profile"
+)
+
+// SettingsExport holds an export of settings.
+type SettingsExport struct {
+	Type Type `json:"type"`
+
+	Config map[string]any `json:"config"`
+}
+
+// SettingsImportRequest is a request to import settings.
+type SettingsImportRequest struct {
+	ImportRequest `json:",inline"`
+
+	// Reset all settings of target before import.
+	// The ImportResult also reacts to this flag and correctly reports whether
+	// any settings would be replaced or deleted.
+	Reset bool `json:"reset"`
+
+	Export *SettingsExport `json:"export"`
+}
+
+func registerSettingsAPI() error {
+	if err := api.RegisterEndpoint(api.Endpoint{
+		Name:        "Export Settings",
+		Description: "Exports settings in a share-able format.",
+		Path:        "sync/settings/export",
+		Read:        api.PermitAdmin,
+		Write:       api.PermitAdmin,
+		Parameters: []api.Parameter{{
+			Method:      http.MethodGet,
+			Field:       "from",
+			Description: "Specify where to export from.",
+		}},
+		BelongsTo: module,
+		DataFunc:  handleExportSettings,
+	}); err != nil {
+		return err
+	}
+
+	if err := api.RegisterEndpoint(api.Endpoint{
+		Name:        "Import Settings",
+		Description: "Imports settings from the share-able format.",
+		Path:        "sync/settings/import",
+		Read:        api.PermitAdmin,
+		Write:       api.PermitAdmin,
+		Parameters: []api.Parameter{{
+			Method:      http.MethodPost,
+			Field:       "to",
+			Description: "Specify where to import to.",
+		}, {
+			Method:      http.MethodPost,
+			Field:       "validate",
+			Description: "Validate only.",
+		}, {
+			Method:      http.MethodPost,
+			Field:       "reset",
+			Description: "Replace all existing settings.",
+		}},
+		BelongsTo:  module,
+		StructFunc: handleImportSettings,
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func handleExportSettings(ar *api.Request) (data []byte, err error) {
+	var request *ExportRequest
+
+	// Get parameters.
+	q := ar.URL.Query()
+	if len(q) > 0 {
+		request = &ExportRequest{
+			From: q.Get("from"),
+		}
+	} else {
+		request = &ExportRequest{}
+		if err := json.Unmarshal(ar.InputData, request); err != nil {
+			return nil, fmt.Errorf("%w: failed to parse export request: %s", ErrExportFailed, err)
+		}
+	}
+
+	// Check parameters.
+	if request.From == "" {
+		return nil, errors.New("missing parameters")
+	}
+
+	// Export.
+	export, err := ExportSettings(request.From)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make some yummy yaml.
+	yamlData, err := yaml.Marshal(export)
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to marshal to yaml: %s", ErrExportFailed, err)
+	}
+
```

Comment:
It would be awesome to check the "Accept" header here and return the expected content type. i.e `application/json` , `text/yaml` or `text/x-yaml`

Reply:
The problem here is with the checksum/signature that we want to add to the export.

I only see a clean way to do this with yaml comments, but not with json...

Reply:
Support for yaml and json is now done.